tests/test_visualization/test_hierarchical_heatmap.py

```diff
@@ -1,98 +1,126 @@
-        expected_mean_intensity = self.adata.to_df().groupby(
-            self.adata.obs['phenotype']
-        ).mean().reset_index()
-        mean_intensity, _ = hierarchical_heatmap(self.adata, 'phenotype')
-        self.assertEqual(mean_intensity.shape, (2, 3))
+
+        # Ensure annotation column is categorical
+        mean_intensity, _, _ = hierarchical_heatmap(self.adata, 'phenotype')
+        self.assertTrue(
+            pd.api.types.is_categorical_dtype(self.adata.obs['phenotype'])
+        )
+
+        # Calculate the expected mean intensity
+        intensities = self.adata.to_df()
+        labels = self.adata.obs['phenotype']
+        expected_mean_intensity = (
+            pd.concat([intensities, labels], axis=1)
+            .groupby('phenotype')
+            .mean()
+        )
+
+        # Ensure the calculated mean intensity matches the expected
+        # mean intensity
         pd.testing.assert_frame_equal(
-            mean_intensity,
-            expected_mean_intensity,
-            check_exact=True
+            mean_intensity, expected_mean_intensity, check_exact=True
+        )
+
+    def test_zscore_normalization(self):
+        """Test z-score normalization."""
+        _, clustergrid, _ = hierarchical_heatmap(
+            self.adata, 'phenotype', z_score=1
+        )
+        self.assertTrue((clustergrid.data2d.mean().round(2) == 0).all())
+
+    def test_clustergrid_attributes(self):
+        """Test if correct clustergrid data attributes are returned."""
+        _, clustergrid, _ = hierarchical_heatmap(self.adata, 'phenotype')
+        self.assertIsInstance(clustergrid, sns.matrix.ClusterGrid)
+
+        # Adjust the expected shape
+        num_unique_phenotypes = self.adata.obs['phenotype'].nunique()
+        self.assertEqual(
+            clustergrid.data2d.shape,
+            (num_unique_phenotypes, self.adata.n_vars)
         )
 
-    def test_matrixplot_attributes(self):
-        """Test if correct matrixplot attributes are returned."""
-        _, matrixplot = hierarchical_heatmap(self.adata, 'phenotype')
-        self.assertEqual(len(matrixplot.var_names), 2)
-        self.assertEqual(len(matrixplot.categories), 2)
-        self.assertListEqual(
-            sorted(list(matrixplot.var_names)),
-            ['feature1', 'feature2']
+    def test_dendrogram_attribute(self):
+        """Test dendrogram attributes."""
+        _, _, dendrogram_data = hierarchical_heatmap(self.adata, 'phenotype')
+
+        # Check if dendrogram_data contains the required linkage keys.
+        self.assertIn('row_linkage', dendrogram_data)
+        self.assertIn('col_linkage', dendrogram_data)
+
+        # Extract linkage matrices and check their structure
+        row_linkage = dendrogram_data['row_linkage']
+        col_linkage = dendrogram_data['col_linkage']
+
+        # Convert linkage arrays to DataFrames and verify their structure
+        expected_columns = ['Node 1', 'Node 2', 'Distance', 'Total Clusters']
```

Comment:
@fangliu117, are these properties of any dendrogram object? If so, we don't need to check if they exist as we did not implement the dendrogram class.

Reply:
Very good point!! Yes, the properties I am testing like the structure of the linkage matrix, are inherent to dendrogram object returned by sns.clustermap, then testing for their existence is redundant, so I have removed those checks, only focusing on our implementation, like testing if the dendrogram data returned from the hierarchical_heatmap function contains certain keys (row_linkage and col_linkage).